internal/powerplatform/services/data_record/api_data_record.go

```diff
@@ -0,0 +1,432 @@
+
+	response, err := client.Api.Execute(ctx, "GET", apiUrl, nil, nil, []int{http.StatusOK}, nil)
+	if err != nil {
+		return ""
+	}
+
+	var mapResponse map[string]interface{}
+	json.Unmarshal(response.BodyAsBytes, &mapResponse)
+
+	oneToMany, _ := mapResponse["OneToManyRelationships"].([]interface{})
+	for _, list := range oneToMany {
+		item := list.(map[string]interface{})
+		if item["ReferencingEntityNavigationPropertyName"] == relationLogicalName {
+			tableName = item["ReferencedEntity"].(string)
+			break
+		}
+		if item["ReferencedEntityNavigationPropertyName"] == relationLogicalName {
+			tableName = item["ReferencingEntity"].(string)
+			break
+		}
+	}
+
+	manyToOne, _ := mapResponse["ManyToOneRelationships"].([]interface{})
+	for _, list := range manyToOne {
+		item := list.(map[string]interface{})
+		if item["ReferencingEntityNavigationPropertyName"] == relationLogicalName {
+			tableName = item["ReferencedEntity"].(string)
+			break
+		}
+		if item["ReferencedEntityNavigationPropertyName"] == relationLogicalName {
+			tableName = item["ReferencingEntity"].(string)
+			break
+		}
+	}
+
+	manyToMany, _ := mapResponse["ManyToManyRelationships"].([]interface{})
+	for _, list := range manyToMany {
+		item := list.(map[string]interface{})
+		if item["Entity1NavigationPropertyName"] == relationLogicalName {
+			tableName = item["Entity1LogicalName"].(string)
+			break
+		}
+		if item["Entity2NavigationPropertyName"] == relationLogicalName {
+			tableName = item["Entity2LogicalName"].(string)
+			break
+		}
+	}
+
+	return tableName
+}
+
+func (client *DataRecordClient) ApplyDataRecord(ctx context.Context, recordId string, environmentId string, tableName string, columns map[string]interface{}) (*DataRecordDto, error) {
+	result := DataRecordDto{}
+
+	environmentUrl, err := client.GetEnvironmentUrlById(ctx, environmentId)
+	if err != nil {
+		return nil, err
+	}
+
+	relations := make(map[string]interface{}, 0)
+
+	for key, value := range columns {
+		if nestedMap, ok := value.(map[string]interface{}); ok {
+			delete(columns, key)
+			if len(nestedMap) > 0 {
+				tableLogicalName, ok := nestedMap["table_logical_name"].(string)
+				if !ok {
+					return nil, fmt.Errorf("table_logical_name field is missing or not a string")
+				}
+				dataRecordId, ok := nestedMap["data_record_id"].(string)
+				if !ok {
+					return nil, fmt.Errorf("data_record_id field is missing or not a string")
+				}
+
+				entityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableLogicalName)
+
+				columns[fmt.Sprintf("%s@odata.bind", key)] = fmt.Sprintf("%s/api/data/%s/%s(%s)", environmentUrl, constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName, dataRecordId)
+			}
+		}
+		if nestedMapList, ok := value.([]interface{}); ok {
+			delete(columns, key)
+			relations[key] = nestedMapList
+		}
+	}
+
+	entityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableName)
```

Comment:
I would refactor this post/patch logic to an external method, its already to much happening in here (over 200 lines of logic)
each step related to API operation could be separated

Reply:
I refactored many parts in functions in this file. no function in api_data_record is more than 80-90 LoC (including whitespace, closing parenthesis in different line, etc.)